src/compile-middleware.ts

```diff
@@ -0,0 +1,108 @@
+const babelLatest = require('babel-preset-latest');
+
+const javaScriptMimeTypes = [
+  'application/javascript',
+  'application/ecmascript',
+  'text/javascript',
+  'text/ecmascript',
+];
+
+const htmlMimeType = 'text/html';
+
+const compileMimeTypes = [
+  htmlMimeType,
+].concat(javaScriptMimeTypes);
+
+function getContentType(response: Response) {
+  const contentTypeHeader = response.getHeader('Content-Type');
+  return contentTypeHeader && parseContentType(contentTypeHeader).type;
+}
+
+function isSuccessful(response: Response) {
+  const statusCode = response.statusCode;
+  return (statusCode >= 200 && statusCode < 300);
+}
+
+export const babelCompile: RequestHandler = transformResponse({
+  shouldTransform(request: Request, response: Response) {
+    return isSuccessful(response) &&
+        compileMimeTypes.indexOf(getContentType(response)) >= 0;
+  },
+
+  transform(request: Request, response: Response, body: string): string {
+    const contentType = getContentType(response);
+    const uaParser = new UAParser(request.headers['user-agent']);
+    const compile = needCompilation(uaParser);
+
+    if (compile) {
+      // TODO(justinfagnani): cache compilation results
+      if (contentType === htmlMimeType) {
+        const document = parse5.parse(body);
+        const scriptTags = dom5.queryAll(document, isInlineJavaScript);
+        for (const scriptTag of scriptTags) {
+          try {
+            const script = dom5.getTextContent(scriptTag);
+            const compiledScriptResult = babelCore.transform(script, {
+              presets: [babelLatest],
+            });
+            dom5.setTextContent(scriptTag, compiledScriptResult.code);
+          } catch (e) {
+            // By not setting textContent we keep the original script, which
+            // might work. We may want to fail the request so a better error
+            // shows up in the network panel of dev tools. If this is the main
+            // page we could also render a message in the browser.
+            console.warn(`Error compiling script in ${request.path}: ${e}`);
+          }
+        }
+        return parse5.serialize(document);
+      } else if (javaScriptMimeTypes.indexOf(contentType) >= 0) {
+        try {
+          const compiledScriptResult = babelCore.transform(body, {
+            presets: [babelLatest],
+          });
+          return compiledScriptResult.code;
+        } catch (e) {
+          console.warn(`Error compiling script in ${request.path}: ${e}`);
+        }
+      }
+    }
+    return body;
+  },
+});
+
+const isInlineJavaScript = dom5.predicates.AND(
+  dom5.predicates.hasTagName('script'),
+  dom5.predicates.NOT(dom5.predicates.hasAttr('src')));
+
+function needCompilation(uaParser: UAParser): boolean {
+  const engine = uaParser.getEngine();
+  const browser = uaParser.getBrowser();
+  const versionSplit = browser.version && browser.version.split('.');
+  const majorVersion = versionSplit ? parseInt(versionSplit[0], 10) : null;
```

Comment:
The tricky thing is that this requires the user to mentally model multi-type comparisons.

I had to spend a little while in a repl to mostly confirm to myself that `null` compares against numbers like it's `0` unless the comparison is `==` or `===`, but I'm still not entirely certain that that's always the case. The nice part about `-1` here is that it's very easy to tell how it will behave in the comparisons that follow.